File: AdventOfCode2019/10/day10.js
const input = require("./input.js");

const getAsteroids = (map) => {
  const asteroids = [];
  for (let y = 0; y < map.length; y++) {
    for (let x = 0; x < map[y].length; x++) {
      if (map[y][x] === "#") {
        asteroids.push([x, y]);
      }
    }
  }

  return asteroids;
};

const isParallel = (v1, v2) => v1[0] * v2[1] === v1[1] * v2[0];
const isSameDirection = (v1, v2) =>
  Math.sign(v1[0]) === Math.sign(v2[0]) && Math.sign(v1[1]) === Math.sign(v2[1]);

const diff = (v1, v2) => [v1[0] - v2[0], v1[1] - v2[1]];

const reduceDirectLineOfSight = (pos, asteroid, visible) => {
  const asteroidRelativeVector = diff(asteroid, pos);
  const onTheSameLine = visible.findIndex((v) => {
    const vector = diff(v, pos);
    return (
      isSameDirection(asteroidRelativeVector, vector) && isParallel(asteroidRelativeVector, vector)
    );
  });
  if (onTheSameLine < 0) {
    visible.push(asteroid);
  } else {
    const onLineAsteroid = visible[onTheSameLine];
    const onLineRelativeVector = diff(onLineAsteroid, pos);
    const k =
      asteroidRelativeVector[0] / onLineRelativeVector[0] ||
      asteroidRelativeVector[1] / onLineRelativeVector[1];
    const isCloser = k < 1;
    if (isCloser) {
      visible[onTheSameLine] = asteroid;
    }
  }
};

const visiblePoints = (asteroids, pos) => {
  const visible = [];
  asteroids.forEach((asteroid) => {
    if (asteroid == pos) {
      return;
    }
    reduceDirectLineOfSight(pos, asteroid, visible);
  });

  return visible;
};

// Part 1
const asteroids = getAsteroids(input.map);
const visible = asteroids.map((x) => visiblePoints(asteroids, x));

const max = visible.reduce(
  (p, c, i) => (p[1].length > c.length ? p : [asteroids[i], c]),
  [[0, 0], []]
);
console.log(max[0], max[1].length);

// Part 2
const dotProduct = (v1, v2) => v1[0] * v2[0] + v1[1] * v2[1];
const vLength = (v1) => Math.sqrt(v1[0] * v1[0] + v1[1] * v1[1]);
const angle = (v1, v2) => Math.acos(dotProduct(v1, v2) / (vLength(v1) * vLength(v2)));
const angleFromUp = (v1) =>
  ((v1[0] >= 0 ? angle([0, -1], v1) : Math.PI + angle([0, 1], v1)) * 180) / Math.PI;

const laserPosition = max[0];
const laserAsteroids = [...max[1]];
laserAsteroids.sort((a, b) =>
  angleFromUp(diff(a, laserPosition)) > angleFromUp(diff(b, laserPosition)) ? 1 : -1
);
const n200 = laserAsteroids[199];
console.log(laserAsteroids.map(x => `(${x[0]},${x[1]})`).join(','));
console.log(n200[0] * 100 + n200[1]);

File: AdventOfCode2019/10/Day10.hs
module Main where

import Data.Function (on)
import Data.List (elemIndices, groupBy, maximumBy, minimumBy, sortOn)
import Data.Ord (comparing)
import qualified Data.Set as Set
import Input (asteroidsMap)

test1 =
  [ ".#..#",
    ".....",
    "#####",
    "....#",
    "...##"
  ]

type Pos = (Int, Int)

findAsteroids :: [[Char]] -> [Pos]
findAsteroids m = foldr rowToVector [] rows
  where
    rows = zip [0 :: Int ..] m
    rowToVector (y, row) acc = [(x, y) | x <- elemIndices '#' row] ++ acc

vSubtract :: Pos -> Pos -> Pos
vSubtract (x1, y1) (x2, y2) = (x1 - x2, y1 - y2)

vAdd :: Pos -> Pos -> Pos
vAdd (x1, y1) (x2, y2) = (x1 + x2, y1 + y2)

vLength :: Floating a => Pos -> a
vLength (x, y) = let p = x * x + y * y in sqrt $ fromIntegral p

vMin v1 v2 = if vLength v1 < vLength v2 then v1 else v2

pseudoNorm :: Pos -> Pos
pseudoNorm (x, y) =
  let vl = gcd x y
   in (x `div` vl, y `div` vl)

isParallel :: Pos -> Pos -> Bool
isParallel (x1, y1) (x2, y2) = x1 * y2 == y1 * x2

isSameDirection :: Pos -> Pos -> Bool
isSameDirection (x1, y1) (x2, y2) = x1 * x2 >= 0 && y1 * y2 >= 0

visibleFromPoint :: Pos -> [Pos] -> Set.Set Pos
visibleFromPoint p asteroids = Set.fromList t
  where
    nonPoint = filter (/= p) asteroids
    vectors = map (`vSubtract` p) nonPoint
    t = foldl addVisibleAsteroid [] vectors
    addVisibleAsteroid (acc@a : as) x =
      if isSameDirection a x && isParallel a x
        then vMin a x : as
        else a : addVisibleAsteroid as x
    addVisibleAsteroid [] x = [x]

maximizeVisiblePoints :: [Pos] -> (Pos, Set.Set Pos)
maximizeVisiblePoints asteroids = maximumBy (comparing (Set.size . snd)) visibleMap
  where
    visibleMap = [(a, visibleFromPoint a asteroids) | a <- asteroids]

part1 = maximizeVisiblePoints . findAsteroids $ asteroidsMap

-- Part 2
laserPosition = fst part1

laserAsteroids = snd part1

dotProduct (x1, y1) (x2, y2) = x1 * x2 + y1 * y2

angle :: Floating a => Pos -> Pos -> a
angle v1 v2 =
  let p = fromIntegral $ dotProduct v1 v2
      cv = p / vLength v1 / vLength v2
   in 180 / pi * acos cv

angleFromUp :: Floating a => Pos -> a
angleFromUp v = if fst v >= 0 then angle (0, -1) v else 180 + angle (0, 1) v

sortFromUp = sortOn angleFromUp

cleanOrder = sortFromUp $ Set.toList laserAsteroids

part2 = laserPosition `vAdd` (cleanOrder !! 199)

main = do
  print $ fst part1
  print $ Set.size $ snd part1
  print part2
